### PR TITLE
feat(sub0): nonfungibles pop-api + pallet implementation

### PR DESCRIPTION
**Pull request by chungquantin**
Nonfungible contracts size: 
```
Original wasm size: 61.5K, Optimized: 21.7K
```
## Description

In an effort to provide a unifying experience for developers to access runtime functionality for non-fungible tokens, APIs compliant to the [PSP34](https://github.com/w3f/PSPs/blob/master/PSPs/psp-34.md#Interfaces) standard will be provided, as well as a set of extra functions (see Management). The interface as defined by this specification will be accessible for smart contracts via:

```rs
use pop_api::nonfungibles::<function>;
```

Note that the lack of version equals latest. As an example, the user could explicitly specify the version as:

```rs
use pop_api::v0::nonfungibles::<function>;
```

The PR includes changes made to the forked pallet-nfts following what has been discussed in the internal spec. Introducing two new pallets; 
- `pallets/nfts`: Changes made to the forked version to add new storage items `AccountBalance` and `Allowances` and update the following logic + unit tests to support the new storage items. 
- `pallets/api/nonfungibles`: Built on top of the `pallet-nfts` to expose READs and DISPATCHs for pop-api library. 

The Nonfungible API provides the contract ability to mint / burn a new item, transfer and approve following the PSP34 Standard.
There are additional features to manage the Nonfungible such as create / destroy a collection, set / clear metadata, set / clear / approve / cancel_approve item attribute. 

## Becnhmarking: 
- [x] Updated pallet-nfts weight file to cover `AccountBalance` read case.
- [ ] Update `approved_transfer`, `transfer` and `cancel_approval` to count `Allowances` storage read.

### CODE REVIEW

**Review by Daanvdplas — COMMENTED** [review]
Amazing work! Especially in the short amount of time. 

I feel there has to be a way to reduce the amount of changes made throughout the repository. E.g. BoundedSlice implements the `PartialEq` trait which we might be able to use and therefore don't have to make the changes you made regarding the `PartialEq`. In addition, I think the changes in the api's pallet mock.rs are not required. This all makes the sub0 branch cleaner in terms of additional changes and reduces the amount of possible errors that can occur.

Not now, but the benchmark and weight results of `destroy` are incorrect. Just leaving this here for later reference, as well as the other comments starting with "Not now,"

For all, inluding the other comments, please be thoughtful on the amount of time spend as we want this to merge asap. It might also have to be rebased after the sponsorship gets merged. 

The integration tests are passing apart from the event assertions for the fungibles, more specifically, in the `transfer_works` and `approve_works` test this event assertion is failing, there might be caused due to the accout mapping, unfortunately I couldn't figure it out. This shouildn't stop merging this PR though.

File: pallets/api/src/mock.rs [hunk 1]
```diff
@@ -1,25 +1,28 @@
...
-pub(crate) const CHARLIE: AccountId = 3;
+pub(crate) const ALICE: u8 = 1;
+pub(crate) const BOB: u8 = 2;
+pub(crate) const CHARLIE: u8 = 3;
 pub(crate) const INIT_AMOUNT: Balance = 100_000_000 * UNIT;
 pub(crate) const UNIT: Balance = 10_000_000_000;
 
 type Block = frame_system::mocking::MockBlock<Test>;
-pub(crate) type AccountId = u64;
+pub(crate) type AccountId = <AccountPublic as IdentifyAccount>::AccountId;
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/api/src/mock.rs:20]: We need this change due to the configuration of pallet-nfts.

File: pallets/nfts/src/features/approvals.rs [hunk 2]
```diff
@@ -173,4 +170,87 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
+		let collection_config = Self::get_collection_config(&collection)?;
+		ensure!(
+			collection_config.is_setting_enabled(CollectionSetting::TransferableItems),
+			Error::<T, I>::ItemsNonTransferable
+		);
+
+		let origin = maybe_check_origin.ok_or(Error::<T, I>::WrongOrigin)?;
+		Allowances::<T, I>::mutate((&collection, &origin, &delegate), |allowance| {
+			*allowance = true;
+		});
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/nfts/src/features/approvals.rs:195]: Sorry but I don't get what you mean here.

File: pallets/api/src/nonfungibles/mod.rs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+			origin: OriginFor<T>,
+			admin: AccountIdOf<T>,
+			config: CreateCollectionConfigFor<T>,
+		) -> DispatchResult {
+			let id = NextCollectionIdOf::<T>::get()
+				.or(T::CollectionId::initial_value())
+				.ok_or(pallet_nfts::Error::<T>::UnknownCollection)?;
+			let creator = ensure_signed(origin.clone())?;
+			let collection_config = CollectionConfig {
+				settings: CollectionSettings::all_enabled(),
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/api/src/nonfungibles/mod.rs:228]: I don't find any issue activating all these settings: https://paritytech.github.io/polkadot-sdk/master/pallet_nfts/enum.CollectionSetting.html

Let me know if you against one of them.

File: pallets/api/src/nonfungibles/mod.rs [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+			NftsOf::<T>::set_collection_max_supply(origin, collection, max_supply)
+		}
+
+		#[pallet::call_index(19)]
+		#[pallet::weight(NftsWeightInfoOf::<T>::mint())]
+		pub fn mint(
+			origin: OriginFor<T>,
+			to: AccountIdOf<T>,
+			collection: CollectionIdOf<T>,
+			item: ItemIdOf<T>,
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/api/src/nonfungibles/mod.rs:352]: Why? There is no `delegate` parameter in this dispatchable function, hence, we can't cancel the collection approval.

File: pallets/api/src/nonfungibles/mod.rs [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+		#[pallet::call_index(7)]
+		#[pallet::weight(NftsWeightInfoOf::<T>::create())]
+		pub fn create(
+			origin: OriginFor<T>,
+			admin: AccountIdOf<T>,
+			config: CreateCollectionConfigFor<T>,
+		) -> DispatchResult {
+			let id = NextCollectionIdOf::<T>::get()
+				.or(T::CollectionId::initial_value())
+				.ok_or(pallet_nfts::Error::<T>::UnknownCollection)?;
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/api/src/nonfungibles/mod.rs:225]: Agree, we can add the collection id as parameter, similar to the fungible

File: pallets/nfts/src/features/create_delete_collection.rs [hunk 6]
```diff
@@ -137,6 +137,11 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 				}
 			}
 
+			// TODO: Do we need another storage item to keep track of number of holders of a
+			// collection
+			let _ =
+				AccountBalance::<T, I>::clear_prefix(collection, collection_details.items, None);
+			let _ = Allowances::<T, I>::clear_prefix((collection,), collection_details.items, None);
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/nfts/src/features/create_delete_collection.rs:144]: https://github.com/r0gue-io/pop-node/issues/335#issuecomment-2472263690

File: pallets/nfts/src/weights.rs [hunk 7]
```diff
@@ -141,6 +124,8 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	/// Proof: `Nfts::CollectionRoleOf` (`max_values`: None, `max_size`: Some(69), added: 2544, mode: `MaxEncodedLen`)
 	/// Storage: `Nfts::Attribute` (r:1001 w:1000)
 	/// Proof: `Nfts::Attribute` (`max_values`: None, `max_size`: Some(479), added: 2954, mode: `MaxEncodedLen`)
+	/// Storage: `Nfts::AccountBalance` (r:1 w:0)
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/nfts/src/weights.rs:127]: https://github.com/r0gue-io/pop-node/issues/335#issuecomment-2472263690

File: pallets/nfts/src/weights.rs [hunk 8]
```diff
@@ -644,39 +641,41 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
 	/// Proof: `Nfts::CollectionConfigOf` (`max_values`: None, `max_size`: Some(73), added: 2548, mode: `MaxEncodedLen`)
 	/// Storage: `Nfts::ItemConfigOf` (r:1 w:0)
 	/// Proof: `Nfts::ItemConfigOf` (`max_values`: None, `max_size`: Some(48), added: 2523, mode: `MaxEncodedLen`)
+	/// Storage: `Nfts::AccountBalance` (r:2 w:2)
```
[the remark this answers is not in the document]
Comment (chungquantin) [reply pallets/nfts/src/weights.rs:644]: https://github.com/r0gue-io/pop-node/issues/335#issuecomment-2472263690

**Pull request merged**